src/libs/elektra/keyset.c

```diff
@@ -1213,29 +1291,43 @@ ssize_t ksCopyInternal (KeySet * ks, size_t to, size_t from)
 elektraCursor ksFindHierarchy (const KeySet * ks, const Key * root, elektraCursor * end)
 {
 	if (ks == NULL || root == NULL) return -1;
+	if (ks->data == NULL)
+	{
+		if (end != NULL)
+		{
+			*end = 0;
+		}
+
+		return 0;
+	}
 
 	ssize_t search = ksSearchInternal (ks, root);
 	size_t it = search < 0 ? -search - 1 : search;
-	if (it == ks->size || keyGetNamespace (root) != keyGetNamespace (ks->array[it]) || keyIsBelowOrSame (root, ks->array[it]) != 1)
+	if (it == ks->data->size || keyGetNamespace (root) != keyGetNamespace (ks->data->array[it]) ||
+	    keyIsBelowOrSame (root, ks->data->array[it]) != 1)
 	{
 		if (end != NULL)
 		{
-			*end = ks->size;
+			*end = ks->data->size;
 		}
-		return ks->size;
+		return ks->data->size;
 	}
 
 	if (end != NULL)
 	{
-		if (root->keyUSize == 3)
-		{
+		struct _KeyName * copy = keyNameCopy (root->keyName);
+		struct _KeyName * old = root->keyName;
+		((Key *) root)->keyName = copy;
+		keyNameRefInc (copy);
```

Comment:
Not needed. Should have been marked as a HACK as well, but what we do here is completely undone before we return. Since we assume a single threaded context, this is okay, because nobody but us sees the modified version.

It's the same as in `ksBelow` where you added "DANGER".

Reply:
Oh I ensure you this is needed ;). This was I bug that I spent way too much time debugging. If a key is in the keyset that is a duplication of `root` this completely breaks. There are some test cases that do exactly that.

Reply:
Hm, okay that would be an issue. But then the detaching should only happen for that one case. If the keys don't share a name, we should not unnecessarily detach since that is a permanent operation.

Reply:
Also couldn't the same issue happen in all the other functions that use this hack? If so, we need to fix those too.

Reply:
We don't detach here. We just temporarily git the `root` key its own copied name. It gets it old name back a few lines after.

Reply:
The other functions don't need it as this behaviour is special to `ksSearchInternal`.

Reply:
Or at least there isn't a single test case that shows this behaviour in other functions.

Reply:
> We don't detach here

Oh yeah, I misremembered. I'll have to look at this and the other functions again... Still seems a bit fishy to me that things are different here

Reply:
Okay, so a few things I found:

1. If `root` itself is in `ks`, I think this code is just broken (even before COW). The `ukey` changes below would always change the name of a key that is in `ks` and therefore skew the result of `ksSearchInternal`.
2. The COW changes, extend this to any COW key that share a name struct with `root`.

In those two cases, do actually need to copy the name. In all other cases, it should be safe to change `ukey` temporarily. Luckily, the two cases are easy to detect, because we already searched for a key in `ks` with the same name as `root` above. Iff `search > 0` then we need the copy.

The exact same logic (1 & 2) applies to `ksBelow` and `ksCut`, but only when `root` uses the cascading namespace. In those functions it's not so easy to detect the two cases. However, we can just make the copy only if `root` is cascading. However, please add a `TODO`, since we plan to prohibit inserting cascading keys into keysets. When that is done, the copy can be removed again.

AFAICT there are no test cases for all that (probably nobody thought of these edge cases before), but it should be easy to add some and confirm my findings.

Reply:
Okay, I've added some tests (the first 12 in `test_ks_cow.c`) for `ksBelow`, `ksCut` and `ksFindHierarchy` for those scenarios:

1. `root` itself is in `ks`
2. A copy of `root` is in `ks`
3. A cascading `root` is in `ks`
4. A copy of a cascading `root` is in `ks`


`ksBelow` has issues with (3), funnywise not with (4).
`ksCut` has no issues.
`ksFindHierarchy` has issues with (3) and (4)

Reply:
Okay so `ksCut` doesn't use `ksFindHierarchy` (even though it could) and uses `elektraKsFindCutpoint` instead. The difference is that `ksFindHierachy` uses `ksSearchInternal` twice, while `elektraKsFindCutpoint` uses only one `ksSearchInternal` and then searches linearly from there. We should definitely update that, because doing two `O(log n)` binary searches is definitely better than one `O(log n)` binary search and then a `O(n)` linear search, even if we sometimes need to make a name copy.

That said, I believe there is still an issue. Try adding e.g. `keyNew ("/root/abc", KEY_END)` to the keyset in the tests. AFAICT this key will never be removed by `ksCut` (or returned by `ksBelow`), which is inconsistent with the fact that `/root` **is** removed. So either we need to also remove cascading keys below `root` or should leave `root` untouched too. I guess this, was always a bug in `ksCut` and is just one more reason to prohibit cascading keys in keysets. But that's a bigger task, so better fix it for now by making a name copy.

Reply:
Okay with the updated tests only the `ksBelow` issues remain.
The `ksFindHierarchy` issues were caused by a misunderstanding of mine of how hierarchies work.

Reply:
Okay so after a lot of confusion and mentally stepping through `ksCut` again and again, I now know why `ksCut` isn't broken. It's a total coincidence. It works, because `KEY_NS_CASCADING` is the first namespace in the order and because it searches linearly from the `ksSearchInternal` result. If it wasn't `ksCut` would return cascading keys for any `root` with a namespace before `KEY_NS_CASCADING`, because of the weird way that `keyIsBelowOrSame` works.

So, two options:
1. Leave `ksCut` as is and add two code comments to the `KEY_NS_CASCADING` section:
     1. `// NOTE: Works only because `KEY_NS_CASCADING` is the first namespace`
     2. `// TODO: Should use ksFindHierarchy like ksBelow`
2. Update `ksCut` to use `ksFindHierarchy` and implement the same fix  as for `ksBelow`

In any case, `ksFindHierarchy` should only do a `keyNameCopy` when actually required, i.e. `if (search > 0)`.

Reply:
>  i.e. if (search > 0).

It's `search >= 0` btw 😉

Reply:
(Didn't read everything in the thread)

> I guess this, was always a bug in ksCut and is just one more reason to prohibit cascading keys in keysets.

Yes, we finally need a decision that KeySets are only allowed to have only keys of the same type. I added to #4645. I think this can directly go to "Decided" state, I don't think there are any argument for allowing such mixed KeySets.

> But that's a bigger task, so better fix it for now by making a name copy.

If there is a decision that we change something, we don't need to fix a bug that gets fixed with implementing the decision anyway (at least during 0.9.*).

> The ksFindHierarchy issues were caused by a misunderstanding of mine of how hierarchies work.

To me it is also not clear how the function is supposed to work, the documentation is poor.

> I think fixing non-broken code is not within the scope of this PR

I fully agree.

> @markus2330 rewriting ksCut may be a topic for FLOSS?

Issues for FLOSS are more-or-less already all taken for this term. C issues in particular are not very popular anyway.

Furthermore, we first need to decide what to do with ksFindHierarchy vs. ksCut. I initially thought this doesn't need a decision but as we are again and again discussing it, we probably should summarize current knowledge about which function to put in the core and which in a helper lib. (Or to keep both if its simply not worth the effort to get rid of one of them.)

Reply:
> (Didn't read everything in the thread)

If you look at the last few comments, you'll see the issue is fixed (also why the thread is "resolved").

A quick summary: We found a flaw in the logic that explains the hacky "just edit `ukey` directly" code in `ksCut`, `ksBelow` and `ksFindHierarchy`. Some of that flaw existed before COW, other parts only exist through COW. In `ksCut` this doesn't result in a bug, because it balances out with the fact that cascading keys are the first in the order of namespaces. In `ksBelow` and `ksFindHierarchy` the issue has been fixed.

The other part about "non-broken code" was about the fact that `ksCut` does not use `ksFindHierarchy` and instead searches linearly for the end of the cut.

> To me it is also not clear how the function is supposed to work, the documentation is poor.

The part that caused confusion was how `Key`s are ordered in a `KeySet` and what "below" means for cascading names. The docs have already been updated in this PR https://github.com/ElektraInitiative/libelektra/commit/e47951a7ec06bbc248f43e2cca01169edf3139a9

> Yes, we finally need a decision that KeySets are only allowed to have only keys of the same type.
>  we first need to decide what to do with ksFindHierarchy vs. ksCut

I'll create those decisions since they either are 100% API or closely related to API questions.

> which function to put in the core and which in a helper lib

Recently, I've come to believe that the core should use indexes (i.e. `elektraCursor`) more, e.g. `ksSearch` not `ksLookup`. Similarly, `ksCut`/`ksBelow` can be replaced with `ksFindHierarchy` and a multi-key equivalent of `ksAtCursor`. I'll go into detail in the decision, but there is already a hint of that idea in the `KeySet` use cases in #4697.